src/clean-core/to_string.cc

```diff
@@ -95,3 +141,834 @@ cc::string cc::to_string(std::byte value)
+{
+    auto const is_neg = value < 0;
+    std::make_unsigned_t<IntType> unsigned_value = is_neg ? -value : value;
+
+    switch (args.sign)
+    {
+    case ' ':
+        ss << (is_neg ? "-" : " ");
+        break;
+    case '-':
+        if (is_neg)
+            ss << "-";
+        break;
+    case '+':
+        ss << (is_neg ? "-" : "+");
+        break;
+    }
+    unsigned_to_string_impl(ss, unsigned_value, args);
+}
+
+template <class FloatType>
+void to_string_float_impl(cc::string_stream& ss, FloatType value, parsed_fmt_args const& args)
+{
+    // note: currently completely relys on std::sprintf
+
+    cc::string_stream sprintf_args;
+    sprintf_args << "%";
+    if (args.sign == ' ')
+        sprintf_args << " ";
+    if (args.sign == '+')
+        sprintf_args << "+";
+    if (args.alternative_mode)
+        sprintf_args << "#";
+    if (args.sign_aware_zero_padding)
+        sprintf_args << "0";
+    if (args.width >= 0)
+        to_string(sprintf_args, args.width, "");
+    if (args.precision >= 0)
+    {
+        sprintf_args << ".";
+        to_string(sprintf_args, args.precision, "");
+    }
+
+    switch (args.type)
+    {
+    case 0: // default, like g but always one digit after decimal point, when fixed point notation is used
+            // note: currently same as g due to limitation of sprintf
+        sprintf_args << "g";
+        break;
+    case 'a': // hexadecimal 0x prefix, 'p' exponent, lowercase letters
+        sprintf_args << "a";
+        break;
+    case 'A': // hexadecimal 0x prefix, 'P' exponent, uppercase letters
+        sprintf_args << "A";
+        break;
+    case 'e': // scientific exponent notation, 'e' exponent
+        sprintf_args << "e";
+        break;
+    case 'E': // scientific exponent notation, 'E' exponent
+        sprintf_args << "E";
+        break;
+    case 'f': // fixed-point
+        sprintf_args << "f";
+        break;
+    case 'F': // fixed-point, but "nan" and "inf" become "NAN" and "INF"
+        sprintf_args << "F";
+        break;
+    case 'g': // general form, precision significant digits and fixed-point or exponent notation depending on magnitute
+        sprintf_args << "g";
+        break;
+    case 'G': // same as 'g' but uppercase version
+        sprintf_args << "G";
+        break;
+    default:
+        CC_ASSERT(cc::always_false<FloatType> && "Invalid format string: Unsupported argument type for float type");
+    }
+
+    // Note: We cannot explicitly convert float as long as we do not run our own float conversion, because sprintf will automatically convert float to double
+    constexpr size_t buffer_size = std::is_same_v<FloatType, float> || std::is_same_v<FloatType, double> ? 31 : 41;
+    char buffer[buffer_size];
+    auto const s = sprintf_args.to_string();
+    auto res = std::snprintf(buffer, sizeof(buffer), sprintf_args.to_string().c_str(), value);
+    CC_ASSERT(res >= 0);
```

Comment:
The error catching switch above could be replaced by passing through `args.type` and testing for snprintf errors like this. For sanity regarding the 31/41 above, this could also assert that `res <= buffer_size`

Reply:
I actually just realized that this is not correct anyways. The buffer size is 31 for `double`, 41 for `long double`.
I am not sure I understand what you mean by passing `args.type`. It does not distinguish `float` from `double`.
I don't think testing` res <= buffer_size` is necessary, as `snprintf` is already passed the buffer size.

Reply:
If snprintf returns a number > buffer size, it would have "wanted" to write more chars than it did, ie. the stack buffer wasn't large enough.

Reply:
With passing args.type i meant just doing `sprintf_args << args.type` without the switch, as this assert would catch formatting errors as well

Reply:
Ah ok 👍

Reply:
Ya I already warned Philip about that switch statement, it's quite useless at the moment :D . I was planning to replace its inner parts by calls to `int_to_hex(...)`, `int_to_octa(...)`, ect.  but rather rushed a `snprintf` based solution to get a working solution. The focus for an initial version should really be API.

Reply:
Wrong switch statement (not the integer one), but same concept.